src/thumbnailer_near_lossless.cc

```diff
@@ -186,32 +199,36 @@ Thumbnailer::Status Thumbnailer::NearLosslessEqual(WebPData* const webp_data) {
     }
 
     if (new_size_psnr.size() == near_ll_frames.size()) {
-      CHECK_THUMBNAILER_STATUS(GenerateAnimationNoBudget(&new_webp_data));
-      if (new_webp_data.size <= byte_budget_) {
-        WebPDataClear(webp_data);
-        *webp_data = new_webp_data;
-        for (int i = 0; i < new_size_psnr.size(); ++i) {
-          const int curr_ind = near_ll_frames[i];
-          frames_[curr_ind].encoded_size = new_size_psnr[i].first;
-          frames_[curr_ind].final_psnr = new_size_psnr[i].second;
-        }
-        final_near_ll = mid_near_lossless;
-        min_near_ll = mid_near_lossless + 1;
-      } else {
-        WebPDataClear(&new_webp_data);
-        max_near_ll = mid_near_lossless - 1;
+      final_near_ll = mid_near_lossless;
+      for (int i = 0; i < new_size_psnr.size(); ++i) {
+        const int curr_ind = near_ll_frames[i];
+        frames_[curr_ind].encoded_size = new_size_psnr[i].first;
+        frames_[curr_ind].final_psnr = new_size_psnr[i].second;
       }
+      min_ind = mid_ind + 1;
+    } else {
+      max_ind = mid_ind - 1;
+    }
+  }
+
+  for (int curr_ind : near_ll_frames) {
+    frames_[curr_ind].config.near_lossless = final_near_ll;
+  }
+
+  if (final_near_ll != 0) {
+    CHECK_THUMBNAILER_STATUS(GenerateAnimationNoBudget(&new_webp_data));
+    if (new_webp_data.size <= byte_budget_) {
+      WebPDataClear(webp_data);
+      *webp_data = new_webp_data;
     } else {
-      max_near_ll = mid_near_lossless - 1;
+      WebPDataClear(&new_webp_data);
```

Comment:
if we discard the last encode attempt, does that mean the final result is not in sync with the settings in frames_? Is that ok?

Reply:
You are right, the final result here must be synchronized with preprocessing 0 if `new_webp_data.size` exceeds the budget, I will fix it!